torch/distributed/fsdp/flat_param.py

```diff
@@ -173,13 +216,13 @@ def init_metadata(
         self._shapes = tuple(shapes)
         self._prefixed_param_names = tuple(prefixed_param_names)
         self._shared_param_infos = tuple(shared_param_infos)
-        self._is_sharded = False
```

Comment:
`FlatParamHandle.uses_sharded_strategy` is equivalent to the old `FlatParameter._is_sharded` because this PR converts the sharding strategy to `NO_SHARD` when the world size is 1.

In other words, the handle's sharding strategy can fully tell us if the handle's `FlatParameter` is ever sharded or not.

Reply:
FlatParamHandle.uses_sharded_strategy can only represent user's config or intention, but _is_sharded is kind of showing the parameter state is sharded or not. I guess it has another way in some other parts I have not reviewed to represent the parameter state is sharded or not?

Reply:
In case there is confusion: `_is_sharded` in the existing code represents if the `FlatParameter` is _ever_ sharded. This is initialized to `False` and set to `True` in `_shard_parameters()` when the sharding strategy is not `NO_SHARD` and the world size is greater than 1:
https://github.com/pytorch/pytorch/blob/e72256604f80e66b6a380479e4b610be19c82e71/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1506-L1509
After that, `_is_sharded` is never changed. Notably, `_is_sharded` does not represent if the `FlatParameter` is _currently_ sharded.

I am probably misunderstanding your point, but given that (1) `_is_sharded` is only set to `True` when not using `NO_SHARD` and when the world size is greater than 1 and (2) in this PR, the sharding strategy is clamped to `NO_SHARD` when the world size is equal to 1, then checking the sharding strategy against `NO_SHARD` should be equivalent to `_is_sharded`.

Reply:
what I meant is '_is_sharded in the existing code represents if the FlatParameter is ever sharded'. after reviewing other parts of the codes, using 'uses_sharded_strategy' sounds good to me, before I worried about whether '_is_sharded' is used to get the self.params_to_flatten. Just checked it uses isInstance(p, FlatParameter) to get self.params_to_flatten